### ISSUE TITLE
Init action for Jupyter and JupyterHub

### ISSUE THREAD

**Issue by nehalecky** [GoogleCloudPlatform/dataproc-initialization-actions#30]
[Jupyter is a widely adopted web-based user interface](http://jupyter.org/) to interactive computing sessions with a variety of kernels. Currently, dataproc init actions contain support for [~~IPython~~ Jupyter Notebook](https://github.com/GoogleCloudPlatform/dataproc-initialization-actions/tree/master/ipython-notebook), however, this does not integrate with recent features [supporting package and environment management in conda](https://github.com/GoogleCloudPlatform/dataproc-initialization-actions/tree/master/conda). 

[JupterHub supports multi-user sessions](https://jupyterhub.readthedocs.org/en/latest/) to Jupyter in a secure and efficient manner.  We should expand on existing init action functionality to also:

- [x] install Jupyter ~~and JupyterHub~~
- [ ] support install via pip or conda
- [x] launch Jupyter~~(Hub)~~ on cluster create
- [x] properly log sessions
- [ ] script / command to remotely start / stop service
- [ ] support use of multiple notebooks per user

**Comment by pquentin** [comment]
@nehalecky This sounds great! The current Jupyter init script leaves a lot to be desired, and needs to be fixed for each Spark release because it specifies the py4j path.

Thanks for working on this.

For a single jupyter instance, I think it's better to launch jupyter from pyspark (using PYSPARK_PYTHON=python and PYSPARK_DRIVER_PYTHON=jupyter) rather than launching spark from jupyter. It makes things much simpler. Not sure if this still applies with JupyterHub.

**Comment by nehalecky** [comment]
@pquentin thanks for the note!

> For a single jupyter instance, I think it's better to launch jupyter from pyspark (using PYSPARK_PYTHON=python and PYSPARK_DRIVER_PYTHON=jupyter) rather than launching spark from jupyter. It makes things much simpler.

This sounds like an interesting approach, and I'd be willing to implement, but I am unfamiliar the param. Could you explain more how it simplifies things (I'd like to understand better)?

> Not sure if this still applies with JupyterHub.

OK, good point. I think whatever mechanism that we associate Jupyter instance with PySpark should be consistent between Jupyter and JupyterHub, do you agree? If so, how could we find out more about this?

Thanks!

**Comment by pquentin** [comment]
This is how we do it:

```
export PYSPARK_PYTHON=python
export PYSPARK_DRIVER_PYTHON=jupyter
export PYSPARK_DRIVER_PYTHON_OPTS='notebook --no-browser --port 8123 --ip="*"'

pyspark \
    --conf spark.executorEnv.PYTHONHASHSEED=1234 \
   ...
    --packages com.databricks:spark-csv_2.11:1.2.0
```

And sure, we need to use the same setup for both Jupyter and JupyterHub.

**Comment by nehalecky** [comment]
closed with #31
